### ISSUE TITLE
AR0002: lot's of false-positive violations

### ISSUE THREAD

**Issue by BrunoJuchli** [BrunoJuchli/StyleCop.CSharp.Async.Rules#2]
A method like

```
    public Task AcquireAsync()
    {
        return _semaphore.WaitAsync();
    }
```

Which returns a task without awaiting it is completely legal in async/await style programming. It's not so common, but maybe still common enough.

Currently the rule AR0002:MethodEndingWithAsyncMustHaveAsyncModifier consider this a violation.
Should the rule be replaced by a rule "Methods ending with Async suffix should either have the async modifier or return a Task/Task<T>"?

In my opinion this would be a better fit.

Please share your opinion.

**Comment by BrunoJuchli** [comment]
Unimplemented methods like:

```
    public Task LogInAsync()
    {
        throw new NotImplementedException();
    }
```

are also affected.

**Comment by Guson** [comment]
Also gets _AR0002_ warnings. I think this should be correct, when using static and generics in an abstract class!?

`protected static async Task<T> SqlProcedureGetByIdAsync(....)`.

Adds this suppression to close the warning:
`[System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.AsyncRules", "AR0002:MethodEndingWithAsyncMustHaveAsyncModifier", Justification = "Uses Async recommend pattern.")]`

Regards
/Jonas

**Comment by BrunoJuchli** [comment]
@Guson 
Thanks, i'll have a look.
For new projects /VS2015 i'd recommend using https://github.com/DotNetAnalyzers/StyleCopAnalyzers

**Comment by BrunoJuchli** [comment]
@Guson 
I've just published version 1.0.31-beta. It should fix your issue with 

```
protected static async Task<T> SqlProcedureGetByIdAsync(....).
```

Can you please verify and give me feedback?

Thanks

**Comment by BrunoJuchli** [comment]
Most of these have now been fixed with 1.0.
Downside however is, that some non-task awaitable return-types will still be seen as a violation and that a custom return type called `Task` - but not being awaitable - will not count as a violation.

If possible i recommend switching to StyleCop.Analyzers which, due to it being based on Rosyln analyzers, has much better verifications.
